### ISSUE TITLE
StateConfiguration does not expose configured State and the StateMachine itself

### ISSUE THREAD

**Issue by beachwalker** [dotnet-state-machine/stateless#45]
The current implementation of the **StateConfiguration does not expose the StateMachine that is currently configured. Not exposing the StateMachine makes it difficult and is a limitation for enriching the fluent interface using Extension Methods** without changing the Stateless library itself (you have to add the current StateMachine as parameter which makes some "noise" that shouldn't be required).

This is an example of an Extension Method if you have a Trigger that flows along several states and should be automatically triggered/fired when reaching a state:

```csharp
public static class StateMachineExtensions
{
	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		IEnumerable<TState> destinations,
		StateMachine<TState, TTrigger> stateMachine)
	{
		return configuration.OnEntry(
			transition =>
				{
					if (!transition.IsReentry && destinations.Contains(transition.Destination))
					{
						stateMachine.Fire(trigger);
					}
				});
	}

	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		TState destination,
		StateMachine<TState, TTrigger> stateMachine)
	{
		return RepeatFireFor(configuration, trigger, stateMachine, destination);
	}

	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		StateMachine<TState, TTrigger> stateMachine,
		params TState[] destinations)
	{
		return RepeatFireFor(configuration, trigger, destinations, stateMachine);
	}
}
```

This results in something like this:

```csharp
stateMachine.Configure(State.SomeState)
	.Permit(Trigger.GotoNext, State.FirstNextState)
	.RepeatFireFor(Trigger.GotoNext, State.SecondNextState, stateMachine)
	.OnEntry(DoSomethingOnEnty);
```

Which could be much cleaner and less error prone (client is forced, could insert the wrong state machine):

```csharp
stateMachine.Configure(State.SomeState)
	.Permit(Trigger.GotoNext, State.FirstNextState)
	.RepeatFireFor(Trigger.GotoNext, State.SecondNextState)
	.OnEntry(DoSomethingOnEnty);
```

... if you are able to write the Extensions this way:
```csharp
public static class StateMachineExtensions
{
	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		IEnumerable<TState> destinations)
	{
		return configuration.OnEntry(
			transition =>
				{
					if (!transition.IsReentry && destinations.Contains(transition.Destination))
					{
                                                // e.g. exposed StateMachine as property here
						configuration.StateMachine.Fire(trigger); 
					}
				});
	}

	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		TState destination)
	{
		return RepeatFireFor(configuration, trigger, destination);
	}

	public static StateMachine<TState, TTrigger>.StateConfiguration RepeatFireFor<TState, TTrigger>(
		this StateMachine<TState, TTrigger>.StateConfiguration configuration,
		TTrigger trigger,
		params TState[] destinations)
	{
		return RepeatFireFor(configuration, trigger, destinations);
	}
}
```

The provision of the StateMachine by the StateConfiguration should be easy, because the StateMachine acts as Factory for the StateMachineConfiguration (even though the current StateConfiguration does not know about the TState/TTrigger generic parameters itself).

Another problem is that you do not **expose the State that is Wrapped into the StateRepresentation's UnderlyingState** that is currently not accessible outside of the StateConfiguration. This way you don't know what state (e.g. enum value) you are currently configuring when creating an Extension Method and you are forced to add some noise to the interface again (add the configured state as param, which the client/user already knows from the factory call .Configure(TState state).

**Comment by nblumhardt** [comment]
Thanks for the suggestions! I'm not sure how much churn this would cause; it doesn't sound like a huge amount. Have you had a shot at it on a private copy? Interested to hear how it came out if so.

**Comment by nblumhardt** [comment]
Closed by #45.
